components/motor/ulnstepper/28byj-48.go

```
@@ -0,0 +1,478 @@
+		if err != nil {
+			return nil, err
+		}
+		m.in1 = in1
+	}
+
+	if mc.Pins.In2 != "" {
+		in2, err := b.GPIOPinByName(mc.Pins.In2)
+		if err != nil {
+			return nil, err
+		}
+		m.in2 = in2
+	}
+
+	if mc.Pins.In3 != "" {
+		in3, err := b.GPIOPinByName(mc.Pins.In3)
+		if err != nil {
+			return nil, err
+		}
+		m.in3 = in3
+	}
+
+	if mc.Pins.In4 != "" {
+		in4, err := b.GPIOPinByName(mc.Pins.In4)
+		if err != nil {
+			return nil, err
+		}
+		m.in4 = in4
+	}
+
+	if err := m.Validate(); err != nil {
+		return nil, err
+	}
+
+	m.startThread(ctx)
+	return m, nil
+}
+
+// struct is named after the controler uln2003.
+type uln2003 struct {
+	theBoard           board.Board
+	rotationPerMinute  float64
+	ticksPerRotation   int
+	in1, in2, in3, in4 board.GPIOPin
+	logger             golog.Logger
+	motorName          string
+
+	// state
+	lock  sync.Mutex
+	opMgr operation.SingleOperationManager
+
+	stepPosition         int64
+	threadStarted        bool
+	targetStepPosition   int64
+	targetStepsPerSecond int64
+	generic.Unimplemented
+}
+
+// validate if this config is valid.
+func (m *uln2003) Validate() error {
+	if m.theBoard == nil {
+		return errors.Errorf("need a board for motor (%s)", m.motorName)
+	}
+
+	if m.rotationPerMinute == 0 {
+		return errors.Errorf("need to set 'rotation_per_minute' for motor (%s)", m.motorName)
+	}
+
+	if m.in1 == nil {
+		return errors.Errorf("need a 'In1' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in2 == nil {
+		return errors.Errorf("need a 'In2' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in3 == nil {
+		return errors.Errorf("need a 'In3' pin for motor (%s)", m.motorName)
+	}
+
+	if m.in4 == nil {
+		return errors.Errorf("need a 'In4' pin for motor (%s)", m.motorName)
+	}
+
+	return nil
+}
+
+// SetPower sets the percentage of power the motor should employ between 0-1.
+func (m *uln2003) SetPower(ctx context.Context, powerPct float64, extra map[string]interface{}) error {
+	if math.Abs(powerPct) <= .0001 {
```

[minor] why is this here? should we just return an error?